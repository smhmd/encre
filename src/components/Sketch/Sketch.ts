import { computed, defineComponent, h, ref } from 'vue';
import { sketchProps, useSketchProvider } from '../../composables';

export default defineComponent({
  name: 'as-sketch',
  props: {
    tag: {
      type: String,
      default: 'div',
    },
    ...sketchProps,
  },
  setup(props, context) {
    const renderer = ref<null | HTMLCanvasElement>(null);
    const { computedViewBox, styles } = useSketchProvider(props);

    return {
      renderer,
      computedViewBox,
      styles,
    };
  },
  methods: {
    genRenderer() {
      return h(
        'svg',
        {
          class: 'ako-svg',
          viewBox: this.computedViewBox,
          xmlns: 'http://www.w3.org/2000/svg',
          ref: 'renderer',
        },
        this.$slots.default && this.$slots.default()
      );
    },
  },
  render() {
    return h(
      this.tag,
      {
        class: 'ako-sketch',
        style: this.styles,
        role: 'Sketch',
      },
      this.genRenderer()
    );
  },
});
