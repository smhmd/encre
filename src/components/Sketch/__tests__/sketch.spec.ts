import { mount, VueWrapper } from '@vue/test-utils';
import Sketch from '../Sketch';

describe('Sketch.ts', () => {
  let mountFunc: (options?: object) => VueWrapper<any>;
  beforeEach(() => {
    mountFunc = (opts = {}) => {
      return mount(Sketch, opts);
    };
  });
  it('should render sketch svg', () => {
    const wrapper = mountFunc({
      props: {
        width: 100,
        height: 100,
      },
      slots: {
        default: 'sketch',
      },
    });
    expect(wrapper.attributes('style')).toContain('width: 100px');
    expect(wrapper.attributes('style')).toContain('height: 100px');
    expect(wrapper.get('svg').attributes('viewBox')).toBe('0 0 100 100');
    expect(wrapper.text()).toBe('sketch');
    expect(wrapper.classes()).toContain('ako-sketch');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
