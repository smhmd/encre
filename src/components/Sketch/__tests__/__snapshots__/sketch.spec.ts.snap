// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sketch.ts should render sketch svg 1`] = `
<div class="ako-sketch"
     style="width: 100px; height: 100px;"
     role="Sketch"
>
  <svg class="ako-svg"
       viewbox="0 0 100 100"
       xmlns="http://www.w3.org/2000/svg"
  >
    sketch
  </svg>
</div>
`;
